db/crud.go

```diff
@@ -842,6 +842,204 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
+				// we previously allocated is unusable now. We have to allocate a new sequence
+				// instead, but we add the unused one(s) to the document so when the changeCache
+				// reads the doc it won't freak out over the break in the sequence numbering.
+				base.InfofCtx(db.Ctx, base.KeyCache, "updateDoc %q: Unused sequence #%d", base.UD(doc.ID), docSequence)
+				unusedSequences = append(unusedSequences, docSequence)
+			}
+
+			for {
+				if docSequence, err = db.sequences.nextSequence(); err != nil {
+					return unusedSequences, err
+				}
+
+				if docSequence > doc.Sequence {
+					break
+				} else {
+					releaseErr := db.sequences.releaseSequence(docSequence)
+					if releaseErr != nil {
+						base.Warnf(base.KeyCRUD, "Error returned when releasing sequence %d. Falling back to skipped sequence handling.  Error:%v", docSequence, err)
+					}
+				}
+			}
+			// Could add a db.Sequences.nextSequenceGreaterThan(doc.Sequence) to push the work down into the sequence allocator
+			//  - sequence allocator is responsible for releasing unused sequences, could optimize to do that in bulk if needed
+		}
+	}
+	doc.Sequence = docSequence
+	doc.UnusedSequences = unusedSequences
+
+	// The server TAP/DCP feed will deduplicate multiple revisions for the same doc if they occur in
+	// the same mutation queue processing window. This results in missing sequences on the change listener.
+	// To account for this, we track the recent sequence numbers for the document.
+	if doc.RecentSequences == nil {
+		doc.RecentSequences = make([]uint64, 0, 1+len(unusedSequences))
+	}
+
+	if len(doc.RecentSequences) >= kMaxRecentSequences {
+		// Prune recent sequences that are earlier than the nextSequence.  The dedup window
+		// on the feed is small - sub-second, so we usually shouldn't care about more than
+		// a few recent sequences.  However, the pruning has some overhead (read lock on nextSequence),
+		// so we're allowing more 'recent sequences' on the doc (20) before attempting pruning
+		stableSequence := db.changeCache.GetStableSequence(doc.ID).Seq
+		count := 0
+		for _, seq := range doc.RecentSequences {
+			// Only remove sequences if they are higher than a sequence that's been seen on the
+			// feed. This is valid across SG nodes (which could each have a different nextSequence),
+			// as the mutations that this node used to rev nextSequence will at some point be delivered
+			// to each node.
+			if seq < stableSequence {
+				count++
+			} else {
+				break
+			}
+		}
+		if count > 0 {
+			doc.RecentSequences = doc.RecentSequences[count:]
+		}
+	}
+
+	// Append current sequence and unused sequences to recent sequence history
+	doc.RecentSequences = append(doc.RecentSequences, unusedSequences...)
+	doc.RecentSequences = append(doc.RecentSequences, docSequence)
+
+	return unusedSequences, nil
+}
+
+func (db *Database) updateExpiryAndTimeSaved(doc *Document, syncExpiry, updatedExpiry *uint32, expiry uint32) (finalExp *uint32) {
```

Comment:
Could be a method on Document instead of Database?

```suggestion
func (doc *Document) updateExpiryAndTimeSaved(syncExpiry, updatedExpiry *uint32, expiry uint32) (finalExp *uint32) {
```